File: java-junit5/src/test/java/org/codecop/Session5_WordCounterRatioTest.java
package org.codecop;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestFactory;

/**
 * Session 5: WordCounterRatioTest - Dynamic/table driven tests. <br />
 * See http://junit.org/junit5/docs/current/user-guide/#writing-tests-dynamic-tests
 */
class Session5_WordCounterRatioTest {

    // TODO Add the needed code/annotations to run this test with all examples.

    @Test
    public void convinceIntelliJThatThisIsValidJUnit5Test() {
        // convince IntelliJ that this is a valid JUnit5 test
    }

    // We want to test more corner cases for the ratio. Here is a table of test cases.
    private static final List<TestCase> TEST_CASES = Arrays.asList(//
            new TestCase("green", "green", 1.0), //
            new TestCase("green bar green", "green", 0.66), //
            new TestCase("green bar green bar", "green", 0.5), //
            new TestCase("green bar green", "bar", 0.33) //
    );

    private static class TestCase {

        private final String sentence;
        private final String word;
        private final double expectedRatio;

        TestCase(String sentence, String word, double expectedRatio) {
            this.sentence = sentence;
            this.word = word;
            this.expectedRatio = expectedRatio;
        }

        String name() {
            return String.format("ratio of '%s' in words '%s' should be %s", word, sentence, expectedRatio);
        }

        void shouldReturnRatioOfGivenWord() {
            WordCounter counter = new WordCounter(sentence);
            assertEquals(expectedRatio, counter.ratioOf(word), 0.01); // keep
        }
    }

    @TestFactory // Test factory will create a sequence of tests.
    List<DynamicTest> createTests() {
        // TODO Take data from TEST_CASES field and convert it to list of DynamicTests.
        // use return Collections.<DynamicTest>emptyList().stream(). //
        // use        collect(Collectors.toList());
        return TEST_CASES.stream(). // drop
                map(testCase -> DynamicTest.dynamicTest(testCase.name(), testCase::shouldReturnRatioOfGivenWord)). // drop
                collect(Collectors.toList()); // drop
    }
}

File: csharp-xunit/WordCountTests/Session5_WordCounterRatioTest.cs
using Org.Codecop.WordCount;
using System.Collections.Generic;
using Xunit;

namespace Org.Codecop.WordCount.Tests
{
    /// <summary>Session 5: WordCounterRatioTest - parameterised/table driven tests.</summary>
    /// <seealso>"TODO"</seealso>
    public class Session5_WordCounterRatioTest
    {
        private static readonly IList<object[]> TestCases = Arrays.AsList(
            new object[] { "green", "green", 1.0 },
            new object[] { "green bar green", "green", 0.66 },
            new object[] { "green bar green bar", "green", 0.5 },
            new object[] { "green bar green", "bar", 0.33 });

        // Mark this test as parameterised
        // TODO Add the needed annotations to run this test with all examples.
        // We want to test more corner cases for the ratio. Here is a table of test cases.
        //
        //
        //
        //
        //
        public static IList<object[]> GetTestCases()
        {
            // This method returns the parameters
            return TestCases;
        }

        private readonly string sentence;

        private readonly string word;

        private readonly double expectedRatio;

        public Session5_WordCounterRatioTest(string sentence, string word, double expectedRratio)
        {
            this.sentence = sentence;
            this.word = word;
            this.expectedRatio = expectedRratio;
        }

        [Fact]
        public virtual void ShouldReturnRatioOfGivenWord()
        {
            WordCounter counter = new WordCounter(sentence);
            Assert.Equal(expectedRatio, counter.RatioOf(word), 0.01); // keep
        }
    }
}
